File: 2018/day02/part1.py
from collections import defaultdict

def freqs(line):
    letter_freqs = defaultdict(int)
    for c in line:
        letter_freqs[c] += 1
    two = False
    three = False
    for v in letter_freqs.values():
        if v == 2:
            two = True
        if v == 3:
            three = True
    return (two, three)


with open('input.txt', 'r') as f:
    twos = 0
    threes = 0
    for line in f:
        two, three = freqs(line)
        if two:
            twos += 1
        if three:
            threes += 1
    print(twos * threes)

File: 2018/day02/part1.clj
(require '[clojure.string :as str])
(require 'clojure.set)

(defn solve [lines]
  (map
    (fn [line] (clojure.set/intersection
        #{2 3}
        (set
          (vals
            (frequencies
              (seq line))))))
    lines))

(defn has-twos-or-threes [line]
  (reduce
    (fn [[two three] v]
      (cond
        (and two three) [true true]
        (= v 2) [true three]
        (= v 3) [two true]
        :else [two three]))
      [false false]
      line))

(println (solve (str/split (slurp "input.txt") #"\n")))
